srtcore/core.cpp

```diff
@@ -8061,21 +8061,6 @@ int srt::CUDT::sendCtrlAck(CPacket& ctrlpkt, int size)
                     uglobal().m_EPoll.update_events(m_SocketID, m_sPollID, SRT_EPOLL_IN, true);
                 }
             }
-#if ENABLE_BONDING
-            if (group_read_seq != SRT_SEQNO_NONE && m_parent->m_GroupOf)
-            {
-                // See above explanation for double-checking
-                ScopedLock glock (uglobal().m_GlobControlLock);
-
-                if (m_parent->m_GroupOf)
-                {
-                    // The current "APP reader" needs to simply decide as to whether
-                    // the next CUDTGroup::recv() call should return with no blocking or not.
-                    // When the group is read-ready, it should update its pollers as it sees fit.
-                    m_parent->m_GroupOf->updateReadState(m_SocketID, group_read_seq);
```

Comment:
I don't understand this, then who will notify app readers to call `Group::recv()`?

Reply:
There's a tough vocabulary on that and I'm unfortunately responsible for all this ;D

Initially I wanted to implement this with a group common buffer, but at that time we had only the receiver buffer for a socket that had lots of builtin problems and could not be reused for the purpose of group reader - this implementation was a complete failure. Therefore I decided to use a different approach - the "app reader", that is, a solution inside the SRT library, but behaving "like it was an application using multiple sockets to read from". This means that every socket uses its own TSBPD read-readiness etc., and the group reader simply extracts the data from the socket using the kinda-wrapped API function for the socket, and sorts these packets out (by verifying whether they are delivered in the sequence number order) AFTER extracting them by the socket's API function.

Now that Max wrote the new receiver buffer, even this poor solution can be improved - instead of manipulating this all through the semi-API calls, the delivery state on every socket can be checked first in the socket's receiver buffer. This PR is to fix some bugs that appeared during the changed reading solution with coexistence of the old rules.

Reply:
Thanks for your reply, I'm looking forward to your group common buffer version, but for the current version in master, here is my understand:
In live mode, group readiness is updated in  tsbpd().
In message mode, group readiness is updated in sendCtrlAck().

Reply:
Once the whole message (all data packets forming it) is in the (new) RCV buffer, you can read it without waiting to be acknowledged.

Reply:
Actually one of the advantage of the receiver buffer is that ACK is no longer a point of read-readiness recognition. This solution only made sense in the stream mode (ever, even in UDT) and should not be used in any other mode. This was because in the stream mode you just have a stream of bytes without boundaries, so the ACK-based triggering had only one motivation: to not trigger read-readiness too often, or just with every single byte of data arriving. But this "too often re-triggering read-readiness" is exclusively a problem in the stream mode and therefore only in the stream mode it should be considered.

Reply:
> Once the whole message (all data packets forming it) is in the (new) RCV buffer, you can read it without waiting to be acknowledged.
Yes, app can ready without waiting ack, but who will notify app to read?

Reply:
In message-mode: when a new packet arrives and the socket isn't ready to read, after insertion of this packet it's checked if this has made at least retrievable message complete (either the first one or any other out-of-order message), and if so, the socket is set read-ready.

In live mode, obviously, TSBPD triggers the read-readiness, updated also by new packet arrival, if needed.

And in stream mode the availability is re-checked in the ACK timer handler (that is, the procedure that timely sends the ACK).

Reply:
> In message-mode: when a new packet arrives and the socket isn't ready to read, after insertion of this packet it's checked if this has made at least retrievable message complete (either the first one or any other out-of-order message), and if so, the socket is set read-ready.

This sounds good, but `processData()` and `insert()` don't update read-readiness now.

Reply:
IIRC, this piece of code could not be removed now.